core/arch/arm/kernel/secure_partition.c

```diff
@@ -217,14 +219,175 @@ static TEE_Result sp_init_uuid(const TEE_UUID *uuid)
 {
 	TEE_Result res = TEE_SUCCESS;
 	struct sp_session *sess = NULL;
+	struct thread_smc_args args = {0};
 
 	res = sp_open_session(&sess,
 			      &open_sp_sessions,
 			      uuid);
 
+	ts_push_current_session(&sess->ts_sess);
+	sp_init_info(to_sp_ctx(sess->ts_sess.ctx), &args);
+	ts_pop_current_session();
+
+	if (sp_enter(&args, sess))
+		return FFA_ABORTED;
+
+	spmc_sp_msg_handler(&args, sess);
+
+	return res;
+}
+
+TEE_Result sp_enter(struct thread_smc_args *args, struct sp_session *sp)
+{
+	TEE_Result res = FFA_OK;
+	struct sp_ctx *ctx = to_sp_ctx(sp->ts_sess.ctx);
+
+	ctx->sp_regs.x[0] = args->a0;
+	ctx->sp_regs.x[1] = args->a1;
+	ctx->sp_regs.x[2] = args->a2;
+	ctx->sp_regs.x[3] = args->a3;
+	ctx->sp_regs.x[4] = args->a4;
+	ctx->sp_regs.x[5] = args->a5;
+	ctx->sp_regs.x[6] = args->a6;
+	ctx->sp_regs.x[7] = args->a7;
+
+	res = sp->ts_sess.ctx->ops->enter_invoke_cmd(&sp->ts_sess, 0);
+
+	args->a0 = ctx->sp_regs.x[0];
+	args->a1 = ctx->sp_regs.x[1];
+	args->a2 = ctx->sp_regs.x[2];
+	args->a3 = ctx->sp_regs.x[3];
+	args->a4 = ctx->sp_regs.x[4];
+	args->a5 = ctx->sp_regs.x[5];
+	args->a6 = ctx->sp_regs.x[6];
+	args->a7 = ctx->sp_regs.x[7];
+
+	return res;
+}
+
+static TEE_Result sp_enter_invoke_cmd(struct ts_session *s,
+				      uint32_t cmd __unused)
+{
+	struct sp_ctx *ctx = to_sp_ctx(s->ctx);
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t exceptions = 0;
+	uint64_t cpsr = 0;
+	struct sp_session *sp_s = to_sp_session(s);
+	struct ts_session *sess = NULL;
+	struct thread_ctx_regs *sp_regs = NULL;
+	uint32_t panicked = false;
+	uint32_t panic_code = 0;
+
+	bm_timestamp();
+
+	sp_regs = &ctx->sp_regs;
+	ts_push_current_session(s);
+
+	cpsr = sp_regs->cpsr;
+	sp_regs->cpsr = read_daif() & (SPSR_64_DAIF_MASK << SPSR_64_DAIF_SHIFT);
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+
+	__thread_enter_user_mode(sp_regs, &panicked, &panic_code);
+	sp_regs->cpsr = cpsr;
+	thread_unmask_exceptions(exceptions);
+
+	thread_user_clear_vfp(&ctx->uctx);
+
+	if (panicked) {
+		DMSG("tee_user_ta_enter: SP panicked with code  %#"PRIx32,
+		     panic_code);
+		abort_print_current_ts();
+
+		sess = ts_pop_current_session();
+		cpu_spin_lock(&sp_s->spinlock);
+		sp_s->state = sp_dead;
+		cpu_spin_unlock(&sp_s->spinlock);
+
+		return TEE_ERROR_TARGET_DEAD;
+	}
+	sess = ts_pop_current_session();
+
+	assert(sess == s);
+
+	bm_timestamp();
 	return res;
 }
 
+static void sp_svc_store_registers(struct thread_svc_regs *regs,
+				   struct thread_ctx_regs *sp_regs)
+{
```

Comment:
With defined(ARM64) both point to `unit64_t` cells with same ordering.
Maybe:
```
	COMPILE_TIME_ASSERT(sizeof(sp_regs->x) == regs->x0);
	memcpy(sp_regs->x, regs->x0, 31 * sizeof(regs->x0));

Reply:
Done